Add poll_recv method to Receiver

jsdw opened the pull request:
This PR Adds a `poll_recv()` method to the `Receiver` type. It returns the same `Result<T,RecvError>` type that the `receiver.recv()` future returns (hence the name).

This method can be used when defining custom streams that internally make use of `async_broadcast` and want to know about whether the `async_broadcast` stream has overflowed or not.

zeenix commented on src/lib.rs:
```diff
@@ -1363,43 +1419,12 @@ impl<T: Clone> Stream for Receiver<T> {
+                // RecvError::Closed is handled upstream, so shouldn't be seen here.
+                None | Some(Err(RecvError::Closed)) => return Poll::Ready(None),
+                // If overflowed, we expect future operations to succeed so try again.
+                Some(Err(RecvError::Overflowed(_))) => continue,
```
I'm not sure I follow how this changes anything. The `Stream` impl is still ignoring the overflow. :thinking:

jsdw replied:
Yup; the idea in this PR was to not change the public Stream impl at all, but to add a way that we could poll the `Receiver<T>`, such that if I'm using `async-broadcast` internally in some struct `Foo`, I can write my own Stream impl for `Foo` which can internally poll the `async_broadcast::Receiver` and take advantage of this overflow information.

niklasad1 replied:
that's is correct but `poll_recv` above is public and one could use that one to implement a custom stream impl with Overflow included...

zeenix replied:
Ah ok, I get it now. Yeah, this is not disruptive at all and mostly just makes most of the inner code public so I don't have any objection to this approach.

zeenix commented on src/lib.rs:
```diff
@@ -1298,6 +1298,62 @@ impl<T: Clone> Receiver<T> {
         }
     }
+
+    /// A low level poll function that returns items of the same format handed back by
```
nitpick: function -> method.

jsdw replied:
Fixed!

niklasad1 commented on src/lib.rs:
```diff
@@ -1298,6 +1298,62 @@ impl<T: Clone> Receiver<T> {
+    pub fn poll_recv(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<T, RecvError>>> {
```
RecvError::Closed is not possible here so would be nice to change the error type to `struct Overflow(pub n)` but this works as well.

```suggestion
    ) -> Poll<Option<Result<T, Overflow>>> {
```

jsdw replied:
Yeahh I wondered about this too. @zeenix what do you reckon? 

1. Stick with `RecvError` from `poll_recv()` even though we'll only use the Overflow variant
2. Add `OverflowError` struct or similar and return `Result<T, OverflowError>`
3. Return  overflow count directly ie `Result<T, u64>` and document this.

I'm leaning to 2 but happy with whatever you prefer :)

I also added a quick test for `poll_recv()` for my own sanity check as much as anything!

zeenix replied:
> Stick with returning `Result<T,RecvError>` from `poll_recv()` even though we'll only use the Overflow variant

Yes, I think this is the best way forward. It's more future-proof this way. 

This reminds me that it would be good to document possible errors from this method in a `Errors` section (as per the conventions).

zeenix replied:
Actually looks like `RecvError` is non_exhaustive so any other variants will require API break anyway. So I'd say we go with @niklasad1's suggestion.

jsdw commented:
I added an example to `poll_recv()`, and added an `OverflowError` so that `poll_recv()` could be more precise rather than returning a `RecvError`.

@zeenix what do you reckon? :)

zeenix commented:
> I added an example to `poll_recv()`

Awesome!

> and added an `OverflowError` so that `poll_recv()` could be more precise rather than returning a `RecvError`.

Is it worth it? Do we envision any other information going in it other than the count? If not, I'd rather keep things simple and just return the integer. If we need an error type, then we just go with option#1.

> > This reminds me that it would be good to document possible errors from this method in a Errors section (as per the conventions).
> 
> I wasn't sure what you meant here offhand (I can't see any such sections in the repo)

I meant adding a `# Error` heading in the docs of the method and listing all possible errors it can throw.

jsdw commented:
> Is it worth it? Do we envision any other information going in it other than the count? If not, I'd rather keep things simple and just return the integer. If we need an error type, then we just go with option#1.

My arguments for keeping `OverflowError` would be:
- The name of the type makes it clear what it is about (ie an error caused because of overflow)
- It acts like the other errors returned in `Result`s in this library, ie it implements `Error` and can be handled accordingly by libs like anyhow and such.

But happy to go either way if you'd prefer otherwise

> I meant adding a # Error heading in the docs of the method and listing all possible errors it can throw.

There are no other Error sections in the library for other functions that return `Results`, but I did what some of those did and added a bit of text about the error condition. Happy to put that into an `# Error` section if you like though?

zeenix commented:
> My arguments for keeping `OverflowError` would be:
> 
>     * The name of the type makes it clear what it is about (ie an error caused because of overflow)

It's clear enough since it's the Err variant of a Result. :) The docs would remove any confusion (if there is still some). :)
 
>     * It acts like the other errors returned in `Result`s throughout this library, ie it implements `Error` and `Display` and can be handled accordingly by libs like anyhow and such.

usize also implements `Display` so that isn't a good argument but `Error` impl is. I'd be fine with `RecvError` here based on this argument.

> There are no other Error sections in the library for other functions that return `Results`,

I didn't realize that but just because we've been lazy so far, doesn't mean we should continue to be so. :) Also in other fallible methods, all variants of the returned result type are possible so it's not as important as when returning an Error and not all variants could be possible (which would be the case for `RecvError` here).

 By "convention", I meant the general conversion in the Rust world (I think it's even recommended in **the** book).

> Happy to put that into an `# Error` section if you like though?

Yes please.

jsdw commented:
Done and done!

zeenix commented:
> Done and done!

You're too efficient. Do it again! :laughing:

zeenix reviewed (changes requested):
Almost done. :)

jsdw commented on src/lib.rs:
```diff
@@ -1298,6 +1298,87 @@ impl<T: Clone> Receiver<T> {
+    /// # Examples
+    ///
+    /// This example shows how the [`Receiver::poll_recv`] method can be used to allow a custom
+    /// stream implementation to internally make use of a [`Receiver`].
```
Ok; I added a bit of text for this too

zeenix reviewed (approved):
:+1:

zeenix commented:
@jsdw now if you could just update the PR description, I can squash merge with that as the commit message. :pray:

niklasad1 commented:
> now if you could just update the PR description, I can squash merge with that as the commit message.

@zeenix good to go? :)
